op-service/sources/eth_client.go

```diff
@@ -352,26 +315,31 @@ func (s *EthClient) PayloadByLabel(ctx context.Context, label eth.BlockLabel) (*
 // It verifies the receipt hash in the block header against the receipt hash of the fetched receipts
 // to ensure that the execution engine did not fail to return any receipts.
 func (s *EthClient) FetchReceipts(ctx context.Context, blockHash common.Hash) (eth.BlockInfo, types.Receipts, error) {
+	receipts, ok := s.recProvider.CachedReceipts(blockHash)
+	if ok {
+		// should always hit the cash
+		info, err := s.InfoByHash(ctx, blockHash)
+		if err != nil {
+			return nil, nil, fmt.Errorf("querying header: %w", err)
+		}
+		return info, receipts, nil
+	}
+
 	info, txs, err := s.InfoAndTxsByHash(ctx, blockHash)
```

Comment:
There's a possible optimization here: If we added a method `InfoAndTxHashesByHash` to the `EthClient`, we could get all the data we need for receipts fetching just with a header query (`eth_getBlockByHash` with 1st param `false`), because header queries already contain all transaction hashes. Note that below we just convert the transactions to a list of hashes.

For this, I think we'd just need to extend the `rpcHeader` type with a field ``Transactions types.Transactions `json:"transactions"` ``

Reply:
Since in pratice, we mostly fetch receipts only if we also need the full blocks, it's probably not worthwhile to invest into this.